src/pip/_internal/utils/direct_url_helpers.py

```diff
@@ -0,0 +1,34 @@
+from pip._internal.models.direct_url import (
+    ArchiveInfo,
+    DirectUrl,
+    DirInfo,
+    VcsInfo,
+)
+
+
+def direct_url_as_pep440_direct_reference(direct_url, name):
+    # type: (DirectUrl, str) -> str
+    """Convert a DirectUrl to a pip requirement string."""
+    direct_url.validate()  # if invalid, this is a pip bug
+    requirement = name + " @ "
+    fragments = []
+    if isinstance(direct_url.info, VcsInfo):
+        requirement += "{}+{}@{}".format(
+            direct_url.info.vcs, direct_url.url, direct_url.info.commit_id
+        )
+    elif isinstance(direct_url.info, ArchiveInfo):
+        requirement += direct_url.url
+        if direct_url.info.hash:
```

Comment:
Actually I'm planning to do that, but I'd rather do it in a followup. It is optional (SHOULD) in PEP 610 and not necessary to implement #609. Also it may require more refactoring of unpack/download to surface the hash information.

Reply:
Actually, I think preserving the hash should be done as part of #4732